### ISSUE TITLE
Permission Errors for environments which do not support OPTIONS requests

### ISSUE THREAD

**Issue by webd-uk** [WordPress/gutenberg#44533]
# Update

It's been identified that the core issue is that some environments don't support `OPTIONS` requests and thus the permissions requests from `@wordpress/core-data` are failing resulting in this error.

This is not specific to the Nav block.

------

### Description

When editing or creating a Navigation Menu the following errors are shown and it doesn't work ...

'You do not have permission to edit this Menu. Any changes made will not be saved.'
'You do not have permission to create Navigation Menus.'

This seems to be caused by an issue with the `/wp-json/wp/v2/navigation?_locale=user` end point on hosting environments that don't have the `OPTIONS` method enabled.

Suggest either changing these requests to user a different method like `GET` or adding a check in "Dashboard - Tools - Health Check" to make sure that the web server has `OPTIONS` enabled as a method.

Oliver

### Step-by-step reproduction instructions

1. Disable `OPTIONS` as a method in [Apache](https://www.ibm.com/support/pages/disabling-options-method-apache-24-http-server) or [Nginx](https://support.plesk.com/hc/en-us/articles/360009961373-How-to-disable-HTTP-OPTIONS-method-in-Nginx-as-a-proxy-in-Plesk-)
2. Enter FSE editor and attempt to edit or create a Navigation Menu

### Screenshots, screen recording, code snippet

More information about this can be found here https://wordpress.org/support/topic/navigation-no-permission-missing-menu/page/4/

### Environment info

- Wordpress v6.0.2
- All Browsers
- All Operating Systems

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by annezazu** [comment]
cc @scruffian in case you have any insight here!

**Comment by talldan** [comment]
`OPTIONS` is used by much of the block editor to check for permissions, so I think the issue will extend beyond the navigation block. 🤔 

Here's the code in question - https://github.com/WordPress/gutenberg/blob/0ab1a64d2dadd663099cf659342580fe14bc0c71/packages/core-data/src/resolvers.js#L310-L314

It might be possible to fallback to a `GET` request if the `OPTIONS` request fails, rather than switching wholesale to `GET`. But then I don't really see why `OPTIONS` is being disallowed. Is there a particular host that does this @webd-uk ?

**Comment by webd-uk** [comment]
Yes, I thought that might be the case because I've seen this before in the Block Widgets interface. That time it was caused by an overzealous security plugin. If you look in the thread you'll see that there are a lot of users with the same problem and so far we have identified one user who didn't have OPTIONS enabled in their Nginx server (I'm not familiar so cannot say if OPTIONS is enabled by default) however the OP's site **does** respond to OPTIONS requests but is getting ...

`[HTTP/2 508 Loop Detected 208ms]`

... errors instead in the editor. I suspect a CDN issue and they have asked their host what the problem is.

Oliver

**Comment by webd-uk** [comment]
Any news on this ... there's [another bunch of people who may be affected](https://wordpress.org/support/topic/you-do-not-have-permission-to-create-navigation-menus-error/).

Oliver

**Comment by talldan** [comment]
@webd-uk It sounds like that's more of an actual permissions issue, but still something worth investigating especially around super admin permissions.

**Comment by getdave** [comment]
Going to ping a few folks here who I know are knowledgable about Core Data including @adamziel and @Mamaduka.

Basically some environments don't support `OPTIONS` and thus permissions requests (which rely on `Headers` from `OPTIONS` requests to determine permission) don't work.

-----

I also removed the Navigation label and renamed the Issue as this is not Navigation specific.

**Comment by getdave** [comment]
> the server can respond with a [405 Method Not Allowed](https://www.rfc-editor.org/rfc/rfc7231#section-6.5.5) status code; in that case, it must give the list of allowed methods in the [Allow response header](https://www.rfc-editor.org/rfc/rfc7231#section-7.4.1);

I wonder if this could help?

**Comment by NerminPh** [comment]
Hello,

I am still encountering the issue with the custom navigation menu block where the message "You do not have permission to edit this Menu. Any changes made will not be saved." appears in the editor area. This issue doesn't affect the frontend, where everything is working well.

Could you please let me know if there are any updates regarding this issue? I would appreciate your help in fixing the appearance of this message in the editor.

Thank you for your assistance.

**Comment by Mamaduka** [comment]
@NerminPh, can you share more details about your environment?

* Any network request failures in the DevTools Network tab? You can sort them by method and see which option requests are failing.
* What WordPress version are you using? WP 6.7 shipped with some improvements for capability checks, and it will avoid making an `OPTIONS` request when possible.
* Are you testing with an Admin role account?

We're working on a long-term solution that will use `HEAD` requests instead of `OPTIONS`, which should avoid similar issues. Track: https://core.trac.wordpress.org/ticket/56481.

## Screenshot

![Image](https://github.com/user-attachments/assets/a7d0c872-68f4-40fd-9df8-d5787aea3d9b)
